exotations/solvers/exotica_control_rrt_solver/include/exotica_control_rrt_solver/control_rrt_solver.h

```diff
@@ -0,0 +1,135 @@
+
+#include <exotica_control_rrt_solver/control_rrt_solver_initializer.h>
+
+// TODO: Remove unused includes
+#include <ompl/base/goals/GoalState.h>
+#include <ompl/base/spaces/SE2StateSpace.h>
+#include <ompl/control/SimpleSetup.h>
+#include <ompl/control/SpaceInformation.h>
+#include <ompl/control/planners/rrt/RRT.h>
+#include <ompl/control/spaces/RealVectorControlSpace.h>
+
+namespace ob = ompl::base;
+namespace oc = ompl::control;
+
+namespace exotica
+{
+class OMPLStatePropagator : public oc::StatePropagator
+{
+public:
+    OMPLStatePropagator(
+        oc::SpaceInformationPtr si,
+        DynamicsSolverPtr dynamics_solver_) : oc::StatePropagator(si), space_(si), dynamics_solver_(dynamics_solver_) {}
+    void propagate(
+        const ob::State *state,
+        const oc::Control *control,
+        const double duration,
+        ob::State *result) const override
+    {
+        double t = timeStep_;
+        space_->copyState(result, state);
+
+        while (t < duration + std::numeric_limits<double>::epsilon())
+        {
+            Integrate(result, control, timeStep_);
+            t += timeStep_;
+        }
+    }
+
+    void setIntegrationTimeStep(double timeStep)
+    {
+        timeStep_ = timeStep;
+    }
+
+    double getIntegrationTimeStep() const
+    {
+        return timeStep_;
+    }
+
+private:
+    double timeStep_ = 0.0;
+    oc::SpaceInformationPtr space_;
+    DynamicsSolverPtr dynamics_solver_;
+
+    void Integrate(ob::State *ob_x, const oc::Control *oc_u, double dt) const
+    {
+        const int NU = dynamics_solver_->get_num_controls();
+        const int NX = dynamics_solver_->get_num_positions() + dynamics_solver_->get_num_velocities();
+
+        double *x = ob_x->as<ob::RealVectorStateSpace::StateType>()->values;
+        double *u = oc_u->as<oc::RealVectorControlSpace::ControlType>()->values;
+
+        Eigen::VectorXd eig_x = Eigen::Map<Eigen::VectorXd>(x, NX);
+        Eigen::VectorXd eig_u = Eigen::Map<Eigen::VectorXd>(u, NU);
+
+        Eigen::VectorXd x_new = dynamics_solver_->Simulate(eig_x, eig_u, dt);
+        std::memcpy(x, x_new.data(), NX * sizeof(double));
+    }
+};
+
+class ControlRRTSolver : public MotionSolver, public Instantiable<ControlRRTSolverInitializer>
+{
+public:
+    ///\brief Solves the problem
+    ///@param solution Returned solution trajectory as a vector of joint configurations.
+    void Solve(Eigen::MatrixXd &solution) override;
+
+    ///\brief Binds the solver to a specific problem which must be pre-initalised
+    ///@param pointer Shared pointer to the motion planning problem
+    ///@return        Successful if the problem is a valid DynamicTimeIndexedProblem
+    void SpecifyProblem(PlanningProblemPtr pointer) override;
+
+private:
+    DynamicTimeIndexedShootingProblemPtr prob_;  ///!< Shared pointer to the planning problem.
+    DynamicsSolverPtr dynamics_solver_;          ///!< Shared pointer to the dynamics solver.
+
+    std::shared_ptr<oc::SimpleSetup> setup_;
```

Comment:
Can you use unique_ptr for all of these?

Reply:
I actually removed these. I was getting `double free or corruption (out)` and I added a bunch of them to be safe but I removed them and now it's fine.

It all comes from `SimpleSetup` taking a bunch of pointers in function args.